Utility classes for writing Java source code from a C++ process

karllessard opened the pull request:
This PR sets up a few tools for writing Java source files from the op generator (or any future generator of Java code written in C++).

They are handling most of the tricky cases known to be encountered when we'll generate the ops, including generics, wildcards, optional inner classes, etc. Note that they focus on simplicity rather than performance or memory optimitizations since they will only be used at compile time.

As stated in a previous PR, java_writer and source_writer are being kept separate since the later might be a good candidate to be move to the core::io layer if required some day.

karllessard committed:
Utility classes for writing Java source code from a C++ process.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

karllessard commented:
@asimshankar , noticing that you have not been assigned this PR yet, in case you missed it... thanks

karllessard committed:
Spliting pull request, retaining only Java definitions into this one.

karllessard commented:
Just a little poke for taking a look at this PR please, thanks!

sb2nov commented:
@asimshankar PTAL

asimshankar reviewed (changes requested):
Sorry for the delay. Haven't looked thorough all of it yet but figured I could send out an initial round of comments/questions.

The PR is pretty big (even conceptually). It might make sense to break it into smaller ones so that we can iterate more quickly. For example:

- One to add all the JavaDefs
- One for SourceWriter
- One for JavaWriter
- One that will hook this up into op_generator

Thoughts?

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,325 @@
+namespace java {
+
+/// \brief An enumeration of different modifiers commonly used in Java
+enum JavaModifier {
```
I juggled a bit with that idea too but I ended up using both the ```java``` namespace and the ```Java*``` prefix where the former classifies any CC code written under the ```//tensorflow/java``` package and the later identifies classes that are directly related to the Java language. Does that make sense to you?

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,325 @@
+class JavaType {
+ public:
+  enum Kind {
+    PRIMITIVE, CLASS, INTERFACE, GENERIC, ANNOTATION, NONE
```
It is only used for empty objects... I tried to focus on simplicity and readability so I tried to avoid dealing with dynamic memory allocation (this code is only run at compile time after all). But looking at it now, the only remaining case of a 'empty' ```JavaType``` is for the return value of a ```JavaMethod```, which should be absent for a constructor. I can probably play a bit around to remove that last case (and the `NONE` value).

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,319 @@
+  template <class TypeScanner>
+  void Scan(TypeScanner* scanner) const;
+
+  /// For sets
```
Agreed, I'll have a look at it

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,319 @@
+///
+/// This class only defines the usage of an annotation in a specific context,
+/// giving optionally a set of attributes to initialize.
+class JavaAnnot {
```
Once again a result from packing up a little bit too much to fit easily in 80 characters, haha! But agreed!

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,319 @@
+  JavaVar() = default;
+  const string& name() const { return name_; }
+  const JavaType& type() const { return type_; }
+  bool periodic() const { return periodic_; }
```
Oops, sorry for that, just a bad translation from french... is it meant for a ```vararg```, I'll change it right away :-/

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,319 @@
+
+// Templates implementation
+
+template <class TypeScanner>
```
In fact, the ```GenericTypeScanner``` should be defined in the ```java_defs.h``` file and can be moved as a private utility in ```java_writer.cc```. There are others implementations (or there will) but are declared by simply using lambdas instead of a named type.

karllessard commented on tensorflow/java/src/gen/cc/java_writer.cc:
```diff
@@ -0,0 +1,262 @@
+  if (!type.params().empty()) {
+    src_writer->Write("<");
+    std::vector<JavaType>::const_iterator it;
+    for (it = type.params().cbegin(); it != type.params().cend(); ++it) {
```
The difference is that ```WriteGenerics(...)``` is used to declare a generic (with all its supertypes in present) while here, we just want to name them. For example, if your have generic ```T``` extending ```Number```, the ```WriteGenerics(...)``` method will output ```<T extends Number>``` while here, you just want to output ```<T>```, like in ```List<T>```.

karllessard commented on tensorflow/java/src/gen/cc/java_writer.cc:
```diff
@@ -0,0 +1,262 @@
+    src_writer->Write((*it)->name());
+    if (!(*it)->supertypes().empty()) {
+      src_writer->Write(" extends ");
+      WriteType((*it)->supertypes().front(), src_writer);
```
Yup, sounds good!

karllessard commented on tensorflow/java/src/gen/cc/source_writer.h:
```diff
@@ -0,0 +1,138 @@
+///
+/// Note: the writer does not acquire the ownership of the file being passed in
+/// parameter.
+class SourceFileWriter : public SourceWriter {
```
In practice, it is this implementation that will be used by the op generator at compile time, the in-memory one (```SourceBufferWriter```) is mainly used for test.

karllessard commented on tensorflow/java/src/gen/cc/source_writer.h:
```diff
@@ -0,0 +1,138 @@
+/// compile-time.
+///
+/// Source writers are language-agnostic and therefore only expose generic
+/// methods common to most languages. Use a decorator class to implement
```
A specific pattern for a wrapper class... I'll clarify this part.

karllessard commented:
Hi @asimshankar , sorry for the delay, I was very busy lately. So as you suggested, I split this PR into smaller ones, starting from the Java definitions. I also incorporated your previous suggestions related to them, please take a look when you'll have a minute, thanks!

asimshankar commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,266 @@
+  const Kind& kind() const { return kind_; }
+  const string& name() const { return name_; }
+  const string& package() const { return package_; }
+  const string& descr() const { return descr_; }
```
Call this `description()` instead?
(Same for other `descr` methods)

asimshankar commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,266 @@
+  const string& descr() const { return descr_; }
+  Type& descr(const string& descr) { descr_ = descr; return *this; }
+  const std::vector<Type>& params() const { return params_; }
+  Type& param(const Type& param) {
```
Let's call this `add_param` (similarly for `add_annotation` and `add_supertype`). That makes it clearer at the call site that these are lists that are added to (as opposed to say `descr()`, which replaces any existing value)

asimshankar commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,266 @@
+  bool IsWildcard() const {
+    return kind_ == GENERIC && name_.empty();
+  }
+  bool operator==(const Type& type) const {
```
Why do we need the operator overloads?

karllessard replied:
Like the comparator, the usage of the == operator will come in a next PR so I'll remove it for now

asimshankar commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,266 @@
+/// \brief A definition of a Java variable
+///
+/// This class defines an instance of a type, which could be documented.
+class Variable {
```
"Variable" here means either an argument to a method or a member variable?

karllessard replied:
Exactly

asimshankar replied:
The name is fine. Perhaps add details in the comments.

asimshankar commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,266 @@
+/// This class defines an instance of a type, which could be documented.
+class Variable {
+ public:
+  static Variable Of(const string& name, const Type& type) {
```
The "Of" here was a little confusing to me. This creates a "variable of type `type` named `name`", right?
Perhaps call it `Variable::Create(const string& name, const Type& type)`?

karllessard replied:
Sorry, that was a terrible name, I was trying to avoid `Create` because I didn't want to prefix all my other factory methods with `Create` (i.e. `CreateIterableOf`, etc.) which makes it too verbose. But thinking of it now, I don't think I'll have to... so yes `Create` it will be :)

asimshankar commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,266 @@
+  const Type& ret_type() const { return ret_type_; }
+  const string& descr() const { return descr_; }
+  Method& descr(const string& descr) { descr_ = descr; return *this; }
+  const string& ret_descr() const { return ret_descr_; }
```
Use `description` instead of `descr` and `return_description()` instead of `ret_descr`?
(And `return_type()` instead of `ret_type()` etc.)

asimshankar commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,266 @@
+  string data_;
+};
+
+// Templates implementation
```
I'd like to see if we can avoid templates. Perhaps remove this from this PR for now and we can discuss options on the follow up PR that will need to introduce some sort of scanner?

karllessard replied:
Ok no problem

karllessard replied:
Ok no problem

asimshankar reviewed (changes requested):
Thanks for making this a smaller PR, was much easier to follow (and nitpick on :)

Some minor things, but this should be good to go soon.

karllessard committed:
Introduce comments from second round of code review

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,266 @@
+  enum Kind {
+    PRIMITIVE, CLASS, INTERFACE, ENUM, GENERIC, ANNOTATION
+  };
+  struct Comparator {
```
Ok I'll remove it for now

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,266 @@
+      return type1.name_ < type2.name_ || type1.package_ < type2.package_;
+    }
+  };
+  static Type Primitive(const string& name) {
```
Sounds good!

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -0,0 +1,266 @@
+  const string& name() const { return name_; }
+  const Type& ret_type() const { return ret_type_; }
+  const string& descr() const { return descr_; }
+  Method& descr(const string& descr) { descr_ = descr; return *this; }
```
Oh I thought this rule was mainly targeting methods arguments and not returned values as well. I can change this but it leads up to some strange syntax in some cases, e.g.
```
type.add_attr(Type::Generic("T"));  // ok
type.add_attr(*(Type::Generic("T").description("any type")));  // mmmh...
```
What do you think?

asimshankar replied:
Yeah, you're right. What you have is fine (returning non-const reference). Sorry for the distraction.

drpngx commented:
@karllessard I am not sure you pushed the comment that addresses @asimshankar comments?

karllessard commented:
@drpngx , you are right, last comments are not pushed yet but I'll do it in the following days, thank you.

karllessard commented:
Ok, that's completed now.

drpngx commented:
```
FAIL: buildifier found errors and/or warnings in above BUILD files.
buildifier suggested the following changes:
112,113d111
<         "src/gen/cc/op_generator.h",
<         "src/gen/cc/op_generator.cc",
114a113,114
>         "src/gen/cc/op_generator.cc",
>         "src/gen/cc/op_generator.h",
Please fix manually or run buildifier <file> to auto-fix.

```

karllessard committed:
Fix .cc/.h order in BUILD file

karllessard commented:
Done, sorry for that.

drpngx commented:
Jenkins, test this please.

asimshankar reviewed (approved):
Thanks! Looking forward to the next ones :)

asimshankar commented:
Jenkins, test this please

asimshankar commented:
Jenkins, test this please

(The previous test failure seems test infrastructure related)

drpngx commented:
FYI I think Jenkins might be stuck today

On Fri, Dec 29, 2017, 10:00 PM Asim Shankar <notifications@github.com>
wrote:

> Jenkins, test this please
>
> (The previous test failure seems test infrastructure related)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/14094#issuecomment-354529507>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AT_SbXSyeu-_lvzQzzJNtqBj0amtZdNcks5tFdFmgaJpZM4QLRYK>
> .
>

asimshankar commented:
The one test failure there is definitely unrelated, so I think this PR is ready to be merged.
